### ISSUE TITLE
Request multiple cores and Solve possible precedence issue with control flow operator

### ISSUE THREAD

**Issue by JiahuaQu** [kircherlab/CADD-scripts#46]
Hi, I found the warning: 
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Because I also found the progress in my own vcf files are very slow, I want to request multiple cores. Do you have any suggestion how to do it?

In addition, there is another waring:
Activating conda environment: ../../../software/CADD/CADD-scripts-master/envs/9f2ffdd26e4a9896cd40b8b91d251805_
Possible precedence issue with control flow operator at /wynton/group/rajkovic/jiahua/software/CADD/CADD-scripts-master/envs/9f2ffdd26e4a9896cd40b8b91d251805_/lib/site_perl/5.26.2/Bio/DB/IndexedBase.pm line 805.
Do you know whether this is important and how we can solve it?
Thank you.

**Comment by aerval** [comment]
Hello,

The script will not run faster if you provide it with multiple cores, as it is not able to use those. If you want to run CADD faster, split your vcf input file into multiple files and run the script separately (in parallel) on those. And if you are not already using those, use the prescored files!

The warning is from Ensembl VEP, not CADD itself. We have found no issue with it yet.

**Comment by JiahuaQu** [comment]
1. Currently, I downloaded and used the "All GRCh38 annotations required for offline installation" file. Does this lead to the slow progress? Did you mean to download "All possible SNVs of GRCh38/hg38 incl. all annotations" + "All gnomad release 3.0 InDels incl. all annotations to initiate a local setup" instead? Where should I save the latter two files and should I unzip them?

2. I installed CADD in a conda environment which should also installed VEP correctly. Do you have any suggestion how to solve it now?

Thank you.

**Comment by JiahuaQu** [comment]
I found that the issue may be raised by the lack of unzip module to unzip .gz files:

**Comment by JiahuaQu** [comment]
Thank you for the above answers.
Still one confusion about: The files are (together with their respective tabix indices) put in the folders no_anno or incl_anno depending on the file under data/prescored/${GENOME_BUILD}_${VERSION}/ and will be automatically detected by the CADD.sh script.

**Comment by JiahuaQu** [comment]
Thank you very much for your answers.
Now I have saved the unzipped prescored files, but encountered another issue:
"Opening /CADD-scripts-master/data/prescored/GRCh38_v1.6/incl_anno/whole_genome_SNVs_inclAnno.tsv.gz..."
I submitted a job in the background but found that the job was frozen and finally died at the step of opening the huge prescored file, whole_genome_SNVs_inclAnno.tsv.gz.

Based on your experience, how much CUP/memory and running time do we need to open that file? Thank you.

**Comment by aerval** [comment]
It is not opening the large .gz file, only loading the tabix index and then doing a single file access at the pinpointed location of the file. You can test it locally by running tabix as described in the first section here: https://cadd.gs.washington.edu/api. Speed-wise, it should be in the 10s of thousands per hour, with the caveat that your local (cluster) file system may slow you down significantly.

**Comment by JiahuaQu** [comment]
Thank you for suggesting testing the retrieval.
Although retrieving SNV CADD scores via tabix was pretty fast, the running program offline was still very slow. The log file still showed the same warning to my last one. I wonder whether the procedure of offline program is different from that of retrieving SNV CADD scores via tabix.

I had already downloaded (1) gnomad.genomes.r3.0.indel_inclAnno.tsv.gz, (2) gnomad.genomes.r3.0.indel_inclAnno.tsv.gz.tbi, (3) whole_genome_SNVs_inclAnno.tsv.gz, and (4) whole_genome_SNVs_inclAnno.tsv.gz.tbi and saved them in my HPC. Because the gnomad.genomes.r3.0.indel_inclAnno.tsv.gz file is smaller than the whole_genome_SNVs_inclAnno.tsv.gz file, the speed and consumed time at the step of opening gnomad.genomes.r3.0.indel_inclAnno.tsv.gz was accepted. However, the step of opening whole_genome_SNVs_inclAnno.tsv.gz was still blocked.

You mentioned that my local (cluster) file system may slow you down significantly. If that is the issue in our case, can you suggest which step or what kind of resource I should ask our HPC administrator to adjust to increase the speed? Thank you.

**Comment by aerval** [comment]
Hello,

The procedure of the offline script is pretty much the same as just using tabix, only via the pytabix library in python.

Please make sure that you have the exactly right files. Check the md5 sums and compare them to those presented on our website. Please make sure that your data is in the right format (i.e chromosome `1` nor `chr1`) as seen in the example data set. And test with small files first, i.e ~5 variants.

Regarding file systems, I am sorry but that goes beyond my understanding. A file system is generally something you set up once when you initialize your system, if your cluster would not work well regarding very large files or frequent random file accesses I do not think you can do much beyond completely setting it up from scratch. However, I do not really believe that this is the issue here, just that it could be possible when doing large scale file accesses. But if it works with some of our files and not with the others, it is more likely a file issue.

**Comment by JiahuaQu** [comment]
Thank you for those suggestions.
I confirmed that files are correct.
Then I tried different strategies to run the program and finally found that only when I extract a very small number of rows of variants to generate a new file to input into the program, CADD can successfully proceed with opening the prescored files and map scores.
It looked that even though I used the offline program installed in my HPC, CADD didn't allow to analyze a complete vcf file which is in a large volume?
